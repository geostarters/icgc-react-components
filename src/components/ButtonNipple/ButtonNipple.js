// @flow

import React from "react";
import PropTypes from "prop-types";
import ReactNipple from "@geostarters/react-nipple";

import IMAGES from "./images/images";
// optional: include the stylesheet somewhere in your app
import styles from "./ButtonNipple.module.css";

export default class ButtonNipple extends React.PureComponent {

	constructor(props: any) {

		super(props);

		this.state = {

			displayNipplePicker: false,

		};

		this.data = null;

	}

	handleClick = () => this.setState({ displayNipplePicker: !this.state.displayNipplePicker });

	handleClose = () => {

		this.setState({ displayNipplePicker: false });
		this.props.handleChangeAngleDef(this.data, this.props.index);

	}

	handleChange = (evt, data) => {

		this.data = data;
		this.props.handleChangeAngle(data, this.props.index);

	}

	radians = a => a * (Math.PI / 180)

	findCoord = (p, d, a) => {

		const b = {x: 0, y: 0};
		a = this.radians(a);
		b.x = p.x - d * Math.cos(a);
		b.y = p.y - d * Math.sin(a);
		return b;

	}

	render() {

		const { angle } = this.props;

		const distance = 50 * angle.exaggeration;
		const degree = angle.degree + 90;

		const frontPosition = this.findCoord({x: 0, y: 0}, distance, degree);

		return (
			<div className={styles.buttonContainer}>
				<div className={ styles.swatch } onClick={ this.handleClick }>
					<div className={ styles.color } style={{backgroundImage: `url(${IMAGES.orientacio})`}}>

					</div>
				</div>
				{ this.state.displayNipplePicker ? <div className={ styles.popover }>
					<div className={ styles.cover } onClick={ this.handleClose }/>
					<div className={ styles.alphaContainer }>
						<div className={ styles.nippleContainer }>
							<ReactNipple
								// supports all nipplejs options
								// see https://github.com/yoannmoinet/nipplejs#options
								options={{
									mode: "static",
									position: { top: "50%", left: "50%" },
									frontPosition,
									color: "#f98f04",
									restJoystick: false,
									restOpacity: 1
								}}
								static={true}
								// any unknown props will be passed to the container element, e.g. 'title', 'style' etc
								style={{
									width: 150,
									height: 150,
									position: "relative",
								// if you pass position: 'relative', you don't need to import the stylesheet
								}}
								// all events supported by nipplejs are available as callbacks
								// see https://github.com/yoannmoinet/nipplejs#start
								onMove={this.handleChange}
							/>
						</div>
					</div>
				</div> : null }
			</div>
		);

	}

}

ButtonNipple.propTypes = {
	handleChangeAngle: PropTypes.func,
	handleChangeAngleDef: PropTypes.func,
	angle: PropTypes.object,
	index: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
};

ButtonNipple.defaultProps = {
	handleChangeAngle: () => {},
	handleChangeAngleDef: () => {},
	angle: {
		degree: 315,
		exaggeration: 1
	},
	index: 0
};
