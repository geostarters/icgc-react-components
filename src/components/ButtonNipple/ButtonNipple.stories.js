//@flow

import React from "react";
import { action } from "@storybook/addon-actions";

import ButtonNipple from "./ButtonNipple";

import "semantic-ui-css/semantic.min.css";

export default {
	title: "ButtonNipple",
	component: ButtonNipple
};

const Template = args => (<ButtonNipple {...args}/>);

export const _ButtonNipple = Template.bind({});
_ButtonNipple.storyName = "ButtonNipple";

export const ButtonNippleAngle = Template.bind({});
ButtonNippleAngle.storyName = "ButtonNipple angle";
ButtonNippleAngle.args = {
	angle: {degree: 150, exaggeration: 0.5},
	handleChangeAngle: action("onChange")
};
