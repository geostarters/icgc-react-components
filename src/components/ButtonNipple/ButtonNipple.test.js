// @flow

import React from "react";
import  { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import ButtonNipple from "./ButtonNipple";
import ReactNipple from "@geostarters/react-nipple";

describe("ButtonNipple Component", () => {

	const buttonNippleProps = {
		handleChangeAngle: jest.fn(),
		handleChangeAngleDef: jest.fn(),
		angle: {
			degree: 315,
			exaggeration: 1
		},
		index: 0
	};

	const component = shallow(<ButtonNipple {...buttonNippleProps}></ButtonNipple>);

	it("renders correct properties", () => {

		expect(toJSON(component)).toMatchSnapshot();

	});

	it("has a element with swatch class", () => {

		expect(component.find(".swatch")).toHaveLength(1);

	});

	it("shows ButtonNipple after click", () => {

		component.find(".swatch").simulate("click");
		expect(component.state("displayNipplePicker")).toBeTruthy();

	});

	it("call handle function after changing angle", () => {

		component.find(ReactNipple).simulate("move", {degree: 180, exaggeration: 0.4});
		expect(buttonNippleProps.handleChangeAngle).toBeCalled();

	});

	it("call handle function after close", () => {

		component.find(".cover").simulate("click", {traget: {}});
		expect(buttonNippleProps.handleChangeAngleDef).toBeCalled();

	});

});
