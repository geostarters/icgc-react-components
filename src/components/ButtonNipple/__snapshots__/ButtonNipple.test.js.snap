// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonNipple Component renders correct properties 1`] = `
<div
  className="buttonContainer"
>
  <div
    className="swatch"
    onClick={[Function]}
  >
    <div
      className="color"
      style={
        Object {
          "backgroundImage": "url(control_orientacio.png)",
        }
      }
    />
  </div>
</div>
`;
